crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,849 @@
+///     char *string = qk_obs_str(obs);
+///     qk_free_obs_str(string);
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+///
+/// The string must not be freed with the normal C free, you must use ``qk_free_obs_str`` to
+/// free the memory consumed by the String. Not calling ``qk_free_obs_str`` will lead to a
+/// memory leak.
+///
+/// Do not change the length of the string after it's returned, although values can be mutated
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_str(obs: *const SparseObservable) -> *mut c_char {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    let string: String = format!("{:?}", obs);
+    CString::new(string).unwrap().into_raw()
+}
+
+/// @ingroup QkSparseObservable
+/// Free a string representation of a ``SparseObservable``.
+///
+///  @param str A pointer to the returned string representation from ``qk_obs_str``
+///
+/// # Safety
+///
+/// Behavior is undefined if ``str`` is not a pointer returned by ``qk_obs_str``
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_free_obs_str(string: *mut c_char) {
+    unsafe {
+        let _ = CString::from_raw(string);
+    }
+}
+
+/// @ingroup QkSparseTerm
+/// Print the sparse term.
+///
+/// @param term A pointer to the term.
+///
+/// @return The function exit code. This is ``>0`` if reading the term failed.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     QkSparseTerm term;
+///     qk_obs_term(obs, 0, &term);
+///     char *string = qk_obsterm_print(&term);
+///     qk_free_obsterm_str(string);
+///
+/// # Safety
+///
+/// Behavior is undefined ``term`` is not a valid, non-null pointer to a ``QkSparseTerm``.
+///
+/// The string must not be freed with the normal C free, you must use ``qk_free_obs_str`` to
+/// free the memory consumed by the String. Not calling ``qk_free_obs_str`` will lead to a
+/// memory leak.
+///
+/// Do not change the length of the string after it's returned, although values can be mutated
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obsterm_str(term: *const CSparseTerm) -> *mut c_char {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let term = unsafe { const_ptr_as_ref(term) };
+
+    let view: SparseTermView = term.try_into().unwrap();
+    let string: String = format!("{:?}", view);
+    CString::new(string).unwrap().into_raw()
+}
+
+/// @ingroup QkSparseObservable
+/// Free a string representation of a ``SparseObservable``.
+///
+///  @param str A pointer to the returned string representation from ``qk_obsterm_str``
+///
+/// # Safety
+///
+/// Behavior is undefined if ``str`` is not a pointer returned by ``qk_obsterm_str``
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_free_obsterm_str(string: *mut c_char) {
```

Comment:
If we're using `obsterm` for the object name of `SparseTerm` (fine by me, just pointing it out), the `free` here is similarly breaking the naming convention.